### PR TITLE
Block Hooks: Apply to Post Content (on frontend and in editor)

### PR DESCRIPTION
**Pull request by ockham**
## Why does this work? And how?

It works the same way as in templates:
- If a post is loaded on the frontend, Block Hooks will be applied to the post content, thus inserting hooked blocks.
- If it is loaded in the editor, the Posts endpoint will inject both the hooked blocks, and the `ignoredHookedBlocks` metadata (into anchor blocks).
- This means that if the post is modified in the editor -- including moving or deleting of hooked blocks -- it will have that `ignoredHookedBlocks` metadata on individual anchor blocks persisted when saving.
- Thus, when loading the post again on the frontend, no hooked blocks will be inserted next to anchor blocks that list them in their `ignoredHookedBlocks`. OTOH, any hooked blocks that were present when saving the post have now simply become part of the markup.

## Testing Instructions, Part One

Start by installing the following plugin, but **do not activate it yet**:

<details>
<summary>Plugin code</summary>

```php
<?php
/**
 * Plugin Name:       Insert Separator blocks Before Headings.
 * Description:       Block Hooks demo plugin that inserts Separator blocks before Heading blocks.
 * Version:           0.1.0
 * Requires at least: 6.7
 */

defined( 'ABSPATH' ) || exit;

function insert_separators_before_headings( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( ! $context instanceof WP_Post ) {
		return $hooked_blocks;
	}

	if ( $anchor_block === 'core/heading' && $position === 'before' ) {
		$hooked_blocks[] = 'core/separator';
	}

	return $hooked_blocks;
}
add_filter( 'hooked_block_types', 'insert_separators_before_headings', 10, 4 );

function set_separator_block_inner_html( $hooked_block, $hooked_block_type, $relative_position, $anchor_block ) {
	if ( $anchor_block['blockName'] === 'core/heading' && 'before' === $relative_position ) {
        $hooked_block['innerContent'] = array( '<hr class="wp-block-separator has-alpha-channel-opacity"/>' );
	}

	return $hooked_block;
}
add_filter( 'hooked_block_core/separator', 'set_separator_block_inner_html', 10, 4 );
```
</details>

- Create a new post that contains a number of headings and paragraphs.
- Save that post, and view it on the frontend. Keep it open in a tab.
- Activate the plugin.
- Go back to the tab with the post (on the frontend). Reload.
- Verify that before each heading, a separator (i.e. a horizontal line) has been inserted.
- Open the post in the editor, and verify that the separators are also present there.
- Modify the separators -- e.g. move one of them around and delete another. Save the post again.
- Reload it on the frontend again. Verify that the changes you made in the editor are respected.

## Screenshots or screencast

block-hooks-post-content

## Testing Instructions, Part Two

_Edit: The behavior covered by the following testing instructions changed after I wrote them, as @sirreal  noticed [here](https://github.com/WordPress/wordpress-develop/pull/7898#issuecomment-2535535295). We ended up deciding that the new behavior -- where hooked blocks would _not_ be inserted if the anchor block was added after the post was added might be preferable, see [this comment](https://github.com/WordPress/gutenberg/pull/67272#issuecomment-2539029074) on the companion GB PR._

- ~Now edit the post, and insert another heading at the bottom. _Note that no separator is inserted at the client side "right away"_ -- the reason is that Block Hooks are almost exclusively handled on the server side. This is one minor UX inconsistency -- the same inconsistency is present in templates.~
- ~Save the post, and reload it on the frontend. Note that a separator has now been inserted before the newly added heading!~ ✨
- ~Finally, reload the post in the editor. Note that the separator is now also present there.~

## Screenshots or screencast, Part Two

block-hooks-post-content-additonal-block

---

Trac ticket: https://core.trac.wordpress.org/ticket/61074

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/default-filters.php
Comment (ockham) [inline src/wp-includes/default-filters.php:770]: I _think_ we also need to set `ignoredHookedBlocks` here.

Comment (ockham) [reply src/wp-includes/default-filters.php:770]: One way of doing this:

```diff
diff --git a/src/wp-includes/blocks.php b/src/wp-includes/blocks.php
index 7ce0356bd8..f07d7b2c93 100644
--- a/src/wp-includes/blocks.php
+++ b/src/wp-includes/blocks.php
@@ -1312,7 +1312,11 @@ function insert_hooked_blocks_into_rest_response( $response, $post ) {
                $content = $response->data['content']['raw'];
        }
 
-       $content = apply_block_hooks_to_content( $content, $post );
+       $content = apply_block_hooks_to_content(
+               $content,
+               $post,
+               'insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata'
+       );
 
        if ( isset( $post_type_to_wrapper_block_mappings[ $post->post_type ] ) ) {
                // Remove mock block wrapper.
```

Note that this also affects `wp_navigation` posts. We need to check (i.e. smoke test, and think through) if this change would affect them adversely (in which case we'd need to add another function for `rest_prepare_post`).

Comment (ockham) [reply src/wp-includes/default-filters.php:770]: Committed in e606abc15e. Let's see how this will work.

File: src/wp-includes/blocks.php
Comment (ockham) [inline src/wp-includes/blocks.php:1323]: We're now inserting hooked blocks _and_ setting `ignoredHookedBlocks` metadata. Previously, we were only inserting hooked blocks (i.e. the default behavior of `apply_block_hooks_to_content` when no third argument is given).

This change also affects the Navigation block. We need to make sure that any hooked blocks inside of the Navigation block -- both as first/last child of `core/navigation` itself, but also `after`/`before` blocks _inside_ of the Navigation block -- still work (both on the frontend and in the editor.

Comment (ockham) [reply src/wp-includes/blocks.php:1323]: To test if the Navigation block still works, we can follow the testing instructions from https://github.com/WordPress/gutenberg/pull/59021.

**Comment by ockham** [comment]
FYI @kmanijak. Maybe you can give this a try and let me know if it works for the purpose you're interested in! 😊

File: src/wp-includes/blocks.php
Comment (ockham) [inline src/wp-includes/blocks.php]: We might need to change the logic here to allow other post types -- pages come to mind, but also other, user-defined CPTs.

Comment (ockham) [reply src/wp-includes/blocks.php]: (All of those probably need to map to `core/post-content`.)

Comment (ockham) [reply src/wp-includes/blocks.php]: ea8012eb1a

File: src/wp-includes/default-filters.php
Comment (ockham) [inline src/wp-includes/default-filters.php:195]: Quoting https://github.com/WordPress/wordpress-develop/pull/7889#discussion_r1858865946:

> `apply_block_hooks_to_content` parses blocks, traverses the block tree (to insert hooked blocks), and reserializes them. Note that it's followed immediately by [`do_blocks`](https://developer.wordpress.org/reference/functions/do_blocks/) (on the line below), which _also_ parses blocks.
> 
> This means that in theory, there's some potential for optimization (parse only once inside of `do_blocks`, and traverse to insert hooked blocks there).
> 
> In practice, I'm not sure if that's viable, or if that might cause some collisions; in particular if `do_blocks` is already used in other contexts where hooked blocks are already inserted (through a different mechanism), e.g. in templates or `wp_navigation` posts.

**Comment by kmanijak** [comment]
> FYI @kmanijak. Maybe you can give this a try and let me know if it works for the purpose you're interested in! 😊

Thank you @ockham for working on this! 🚀 I gave it a try and at first glance it works as expected! I can achieve the goals raised in scope of the [discussion in WooCommerce](https://github.com/woocommerce/woocommerce/discussions/52915). I'll play with it more tomorrow to check if I see any problems.

One potential feature I can already see (but obviously not the scope for this PR!) is the ability for more strict selectors. For example: `$anchor_block` as `core/post-title` but only within Query Loop, not _any_ `core/post-title`, including the main post title. This very specific example can be workaround by checking `if ($context instanceof WP_Post) {` but wanted to give some generic example.

Anyway, looking good, thank you! 🙌

**Comment by leewillis77** [comment]
Just following up on @kmanijak's comments. We flagged this as an issue to the Woo team, specifically about hooking blocks into Woo's Product Collection block when it was included in pages. Unfortunately, the patch here doesn't appear to resolve the issues we were seeing (we're unable to add hooked blocks inside the Product Collection block). 

It's possible that the issue is Woo-specific (or on our side), but I just wanted to drop my observations here in case it is fundamental to the solution. 
https://github.com/woocommerce/woocommerce/discussions/44776#discussioncomment-11414874

**Comment by leewillis77** [comment]
I'm testing with the WP 6.7.1 with the patch from this issue. Are there other dependent changes to make this work (ie would I have to start with WP trunk?)

**Comment by ockham** [comment]
Thanks a lot for testing, @leewillis77! I'm a bit puzzled as to why this wouldn't work if patched against 6.7.1; there shouldn't really be any dependent changes required for this.

Can you maybe try with the [playground link](https://playground.wordpress.net/wordpress.html?pr=7898) (from [this comment](https://github.com/WordPress/wordpress-develop/pull/7898#issuecomment-2501399288))?

**Comment by ockham** [comment]
For reviewers: @kmanijak [has kindly tested this](https://github.com/woocommerce/woocommerce/discussions/44776#discussioncomment-11415607) using the Woo Product Collection block and my [Like Button block](https://github.com/ockham/like-button/releases/latest) in a number of different scenarios 🙂

**Comment by leewillis77** [comment]
@ockham @kmanijak I've re-tested this today (fresh install, re-patch etc.), and can confirm that block hooks *are* fired correctly on posts/pages. No idea why we couldn't get this working last week.

However, trying our actual implementation rather than just artificial tests I don't think this is going to give us what we actually need (what was originally raised on the WooCommerce discussion) due to the lack of contextual information (as @kmanijak alluded to in https://github.com/WordPress/wordpress-develop/pull/7898#issuecomment-2506684117). 

Ideally what we need in block hooks when we're deciding whether or not to hook a block in or not is the context of where it is. This is the same issue I originally raised in https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1875449258. It's particularly bad here as the `$context` isn't even the top-level block, but a `WP_Post` object so we don't even have any idea what top-level block is being rendered so that we can make an informed decision about whether we should be hooking in or not.

File: src/wp-includes/blocks.php
Comment (gziolo) [inline src/wp-includes/blocks.php]: `get_post` can still return `null`. It’s fine type wise, but maybe the documentation should state that it will try to get the current post.

File: src/wp-includes/blocks.php
Comment (gziolo) [inline src/wp-includes/blocks.php:1294]: What’s the reasoning here?

File: src/wp-includes/blocks.php
Comment (gziolo) [inline src/wp-includes/blocks.php:1307]: What about synced patterns `wp/block`? Should it be extensible? These post types are special do maybe it’s fine as is but a filter here could be useful.

File: src/wp-includes/default-filters.php
Comment (gziolo) [inline src/wp-includes/default-filters.php:768]: Do you remember why the line above was needed? ` update_ignored_hooked_blocks_postmeta` could have something with the first and last child. Maybe same handling is missing for post and pages? More broadly how the same functionality could work with CPTs?

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1294]: A unit test was failing (see [e.g.](https://github.com/WordPress/wordpress-develop/actions/runs/12073501688/job/33669751861#step:14:869)).

These fields can be set, but their value(s) could be the empty string. We didn't really handle that case specifically, and it seems that some part of the Block Hooks logic would later end up setting the `raw` field to `false` if that was the case (which broke the unit test, as it expected it to still be equal to the empty string).

The best way to avoid this (IMO) is to skip the Block Hooks algorithm altogether if there's not content to apply it to 😄 Hence the change.

Comment (gziolo) [reply src/wp-includes/blocks.php:1294]: Makes perfect sense 👍🏻

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php]: I've covered this [here](https://github.com/WordPress/wordpress-develop/pull/7898/files#diff-8c99af92e4ec0fdb307ddd9b42be1e1ef1efe4a9f31287c23f346244dddd1ce9R1047-R1048):

> If set to `null`, the current post is used. Default: `null`.

To make it more accurate, I can change it to something like

> If set to `null`, `get_post()` will be called to use the current post as context. Default: `null`.

WDYT?

Additionally, I can change the `@since` line to

```suggestion
 * @since 6.8.0 Have the `$context` parameter default to `null`, in which case `get_post()` will be called to use the current post as context.
```

Comment (gziolo) [reply src/wp-includes/blocks.php]: Works for me 👍🏻

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1307]: Ah, good point. I guess the `wp_block` post type should map to `core/block`.

A filter might make sense here 👍 Something to map post types to block types. (I wonder if we already have a function or data structure to do that somewhere 🤔 )

File: src/wp-includes/default-filters.php
Comment (ockham) [inline src/wp-includes/default-filters.php:770]: This follows the precedent of `rest_prepare_wp_navigation` below. However, it has the unfortunate side effect that Block Hooks aren't automatically applied to user-defined CPTs; those would need to be added manually (via `add_filter( 'rest_prepare_my_cpt', ... )`). (AFAIK, there's no "generic" `rest_prepare` filter that could cover all post types.)

I wonder if we should run `insert_hooked_blocks_into_rest_response` from inside the Posts controller instead, in order to insert hooked blocks into every CPT that has `show_in_rest` set to `true` 🤔 cc/ @gziolo

File: src/wp-includes/default-filters.php
Comment (ockham) [reply src/wp-includes/default-filters.php:768]: Good spot 👍 Yeah, `update_ignored_hooked_blocks_postmeta` is almost certainly about first/last child insertion. I need to test if that's currently working with posts and pages. If it isn't, I'll probably have to add it for those post types.

Comment (ockham) [reply src/wp-includes/default-filters.php:768]: > I need to test if that's currently working with posts and pages. If it isn't, I'll probably have to add it for those post types.

Ah, so this is tricky. Since `update_ignored_hooked_blocks_postmeta` is only needed to set `ignoredHookedBlocks` metadata, I expected that a hooked block that's inserted as the Post Content block's last child would at least be rendered on the frontend.

However, it isn't! The reason is the same as for the Navigation and Template Part blocks: these are rendered by separate logic (in the respective blocks' dynamic HTML). For Navigation and Template Part, we changed those block's `render()` methods accordingly (e.g. by making sure the Template Part uses the correct functions that will insert hooked blocks into the underlying `WP_Template` object, https://github.com/WordPress/gutenberg/pull/55811).

In this PR OTOH, we're using the `the_content` filter for hooked blocks insertion -- which works fine for all hooked blocks that are supposed to be inserted _inside_ post content, but not for the ones that look for a `core/post-content` block as their anchor block -- because it simply isn't there.

One possible fix to enable this at least on the frontend is to move the logic into the Post Content block's render method. However, this might have the downside that it will only insert hooked blocks when used with a block theme -- whereas the current solution (via `the_content`) should also work if a classic theme is used 🤔

Comment (ockham) [reply src/wp-includes/default-filters.php:768]: There's also the somewhat philosophical question if a block that's hooked as `core/post-content`'s `lastChild` should be inserted if we're using a classic theme -- where there is no `core/post-content` block, strictly speaking 😅

Comment (ockham) [reply src/wp-includes/default-filters.php:768]: This seems to work (to get a block hooked has `core/post-content`'s `last_child` to be rendered on the frontend):

```diff
diff --git a/src/wp-includes/blocks/post-content.php b/src/wp-includes/blocks/post-content.php
index 25be880cc4..97f030095a 100644
--- a/src/wp-includes/blocks/post-content.php
+++ b/src/wp-includes/blocks/post-content.php
@@ -46,10 +46,25 @@ function render_block_core_post_content( $attributes, $content, $block ) {
                $content .= wp_link_pages( array( 'echo' => 0 ) );
        }
 
+       // Wrap in Post Content block so the Block Hooks algorithm can insert blocks
+       // that are hooked as first or last child of `core/post-content`.
+       $content = get_comment_delimited_block_content(
+               'core/post-content',
+               $attributes, // TODO: Merge ignoredHookedBlocks from post meta.
+               $content
+       );
+
+       // We need to remove the `core/post-content` block wrapper after the Block Hooks algorithm,
+       // but before `do_blocks` runs, as it would otherwise attempt to render the same block again --
+       // thus recursing infinitely.
+       add_filter( 'the_content', 'remove_serialized_parent_block', 8 );
+
        /** This filter is documented in wp-includes/post-template.php */
        $content = apply_filters( 'the_content', str_replace( ']]>', ']]&gt;', $content ) );
        unset( $seen_ids[ $post_id ] );
 
+       remove_filter( 'the_content', 'remove_serialized_parent_block', 8 );
+
        if ( empty( $content ) ) {
                return '';
        }
```

This means I'll have to cover this on the GB side.

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1323]: Turns out there was an unrelated issue with this. I've filed a fix: https://github.com/WordPress/wordpress-develop/pull/7941

File: src/wp-includes/default-filters.php
Comment (ockham) [reply src/wp-includes/default-filters.php:768]: (Updated https://github.com/WordPress/gutenberg/pull/67272 to include the required changes.)

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php]: Fixed in https://github.com/WordPress/wordpress-develop/pull/7898/commits/2a1f4354e5e03cb2faa722cc66d903197e04cc93 and https://github.com/WordPress/wordpress-develop/pull/7898/commits/ff37889f604772494f8d4aeb6cb9c64a1f47606e.

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1323]: I've committed #7941 to Core [`[62639]`](https://core.trac.wordpress.org/ticket/62639), and rebased this PR to include the fix.

**Comment by ockham** [comment]
@gziolo Since insertion as the Post Content block's first or last child required changes to that block's PHP, I've updated the [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/67272) -- it also includes the required compat layer changes backported from this PR. Would you mind reviewing giving the GB PR a look?

There are a few open questions on this PR:
- [Add filter to allow different post type/block type mappings?](https://github.com/WordPress/wordpress-develop/pull/7898#discussion_r1867077508)
- [Use `rest_prepare_` hook to insert hooked blocks into REST API response, or add inline in controller?](https://github.com/WordPress/wordpress-develop/pull/7898#discussion_r1867470796)

However, I don't think that those should be blockers for the Gutenberg PR. Both of them deserve some more consideration, but we might do that as part of this PR -- possibly even after the GB PR has landed 😊

**Comment by sirreal** [comment]
I was following the testing instructions from the description and when I add more headings the separator is not being added 😕 

> Testing Instructions, Part Two
> …
> - Save the post, and reload it on the frontend. Note that a separator has now been inserted before the newly added heading! ✨
> - Finally, reload the post in the editor. Note that the separator is now also present there.

**Comment by sirreal** [comment]
> I was following the testing instructions from the description and when I add more headings the separator is not being added 😕

When I add more headers they all look like this, which I believe is _not_ intended according to the testing instructions:

```html
<!-- wp:heading {"level":1,"metadata":{"ignoredHookedBlocks":["core/separator"]}} -->
<h1 class="wp-block-heading">text</h1>
<!-- /wp:heading -->
```

**Comment by ockham** [comment]
@sirreal Good spot! Turns out @gziolo [noticed the same thing](https://github.com/WordPress/gutenberg/pull/67272#issuecomment-2536039160) over at the GB PR. I [replied there](https://github.com/WordPress/gutenberg/pull/67272#issuecomment-2539029074). The _tl;dr_ is 

> Apologies, I think that this change in behavior is a side effect of some code I pushed after I wrote the testing instructions.
> It's actually not bad in terms of UX -- a bit more predictable maybe. I'm happy to just consider it a happy accident and land it as-is.

So we might keep it this way. I'll add a note to the testing instructions.

File: src/wp-includes/blocks.php
Comment (sirreal) [inline src/wp-includes/blocks.php:1059]: Could this _still_ be null and would that be a problem? `null` is in the return type of [`get_post()`](https://developer.wordpress.org/reference/functions/get_post/).

File: src/wp-includes/blocks.php
Comment (sirreal) [inline src/wp-includes/blocks.php]: This is all about calling `apply_block_hooks_to_content` (above) with a different third argument, is that right? It's confusing to see that it's called above then the filter that would call it here is removed, but it makes sense if it's about that third argument.

The comments could be a bit clearer.

```suggestion
	// `apply_block_hooks_to_content` is called above. Ensure it is not called again as a filter.
	$priority = has_filter( 'the_content', 'apply_block_hooks_to_content' );
	if ( false !== $priority ) {
		remove_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
	}

	/** This filter is documented in wp-includes/post-template.php */
	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );

	// Restore the filter if it was set initially.
	if ( false !== $priority ) {
		add_filter( 'the_content', 'apply_block_hooks_to_content', $priority );
	}
```

**Review by sirreal — APPROVED** [review]
This tests well for me. I'm not well versed in block hooks but the changes seem reasonable. I've left a few notes, but I think this can be merged if you're confident.

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php]: > This is all about calling `apply_block_hooks_to_content` (above) with a different third argument, is that right? It's confusing to see that it's called above then the filter that would call it here is removed, but it makes sense if it's about that third argument.

Yeah, that's pretty much the reason. We're calling `apply_block_hooks_to_content` here with the `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` callback, as we'd like the block markup returned by the REST API for use in the editor to contain both hooked blocks, _and_ the `metadata.ignoredHookedBlocks` attributes.

After that step, we apply the `the_content` filter to the blocks markup (which at this also includes hooked blocks and `ignoredHookedBlocks` metadata) to get the "rendered" version (which -- among other things -- runs `do_blocks` to expand block delimiters into "rendered" markup, by running individual blocks' `render_block` methods etc).

OTOH, we have "globally" added `apply_block_hooks_to_content` to the `the_content` filter. (This is necessary for hooked blocks to be inserted into Post Content on the frontend -- including sites that use a Classic Theme rather than a Block Theme.) By default, `apply_block_hooks_to_content` uses `insert_hooked_blocks` as its callback, which is fine for the frontend (we don't need to set `ignoredHookedBlocks` metadata there).

But since we've already applied Block Hooks earlier in this process, we now remove the filter and re-add it only after the (other) `the_content` filter have been applied.

Thinking about it now, there _might_ be alternatives, such as using `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` for the frontend (i.e. in `the_content`) as well; but it seems kinda wasteful to set `ignoredHookedBlocks` metadata when we don't even need it. So I'm inclined to leave as-is for now, but open to reconsidering and tweaking this some more. LMK if you have any other thoughts or suggestions, maybe I'm not seeing all possible alternatives.
 
> The comments could be a bit clearer.

Those make a lot of sense, I'll apply them. Thank you!

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1059]: Hmm, good question 🤔 The Block Hooks algorithm should work even with no `$context` specified -- it started out as an optional argument that filters such as `hooked_block_types` could use to conditionally insert hooked blocks based on context.

I _think_ that that should still be the case, but I might wanna double-check 👍

Comment (ockham) [reply src/wp-includes/blocks.php:1059]: Seems fine. I've added test coverage in a7693b07ac.

File: src/wp-includes/blocks.php
Comment (sirreal) [reply src/wp-includes/blocks.php]: > Thinking about it now, there _might_ be alternatives, such as using `insert_hooked_blocks_and_set_ignored_hooked_blocks_metadata` for the frontend (i.e. in the_content) as well; but it seems kinda wasteful to set `ignoredHookedBlocks` metadata when we don't even need it

There is some appeal in removing the filter swapping implementation that's here. What's involved in the wastful processing, is it parsing the block with the HTML API unnecessarily or something along those lines?

File: src/wp-includes/blocks.php
Comment (ockham) [reply src/wp-includes/blocks.php:1307]: I've finally tested this with a synced pattern, and it's currently _not_ inserting hooked blocks.

Simply adding `elseif ( 'wp_block' === $post->post_type ) { $wrapper_block_type = 'core/block'; }` didn't do the trick, so I'll have to look into this a bit more. However, since this PR has been open for a while and seems to be ready otherwise, I'll merge it and tackle synced patterns in a follow-up 😊

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/59523.

I'll file some follow-up tickets/PRs to cover some open questions:
- [Enable for synced patterns (`core/block`)](https://github.com/WordPress/wordpress-develop/pull/7898#discussion_r1867077508)
- [Potentially add filter to extend set of block types and post types that Block Hooks are applied to](https://github.com/WordPress/wordpress-develop/pull/7898#discussion_r1867464844)
- [Potentially inline code into Posts controller to cover "all" post types (rather than manually adding the filter to individual `rest_prepare_*` filters)](https://github.com/WordPress/wordpress-develop/pull/7898/files#r1867470796)

File: src/wp-includes/default-filters.php
Comment (gziolo) [reply src/wp-includes/default-filters.php:770]: It would be worthwhile to explore supporting other types than `post`, `page`, and `wp_navigation`.

File: src/wp-includes/blocks.php
Comment (gziolo) [reply src/wp-includes/blocks.php:1307]: That's fair. It was also discussed in https://github.com/WordPress/wordpress-develop/pull/7898/files#r1867470796 to add support for every post type. The challenge with Patterns is that they are handled out of the box if they come from files bundled from the theme or when registered with PHP. The last missing puzzle would be a synced version of the pattern.

Comment (ockham) [reply src/wp-includes/blocks.php:1307]: Update: Support for synced patterns was implemented in https://github.com/WordPress/gutenberg/pull/68058 and https://github.com/WordPress/wordpress-develop/pull/8015, respectively.

I've filed a Trac ticket for other CPTs (and filterable post type -- block type mappings): https://core.trac.wordpress.org/ticket/62715

File: src/wp-includes/default-filters.php
Comment (ockham) [reply src/wp-includes/default-filters.php:770]: I've filed https://core.trac.wordpress.org/ticket/62715 to keep track of this.

**Pull request closed without merging**
